src/userobject/ElasticRecoilCrossSectionUserObject.C

```diff
@@ -0,0 +1,205 @@
+  : GeneralUserObject(parameters),
+    _neutron_spectrum(getFunction("neutron_spectrum")),
+    _scattering_law(getFunction("scattering_law")),
+    _simple_erxs(getFunction("simple_erxs")),
+
+    _atomic_mass(getParam<std::vector<Real>>("atomic_mass")),
+    _isotope_type(getParam<std::vector<unsigned int>>("isotope_type")),
+    _number_isotope(_isotope_type.size()),
+    _neutron_energy_limits(getParam<std::vector<Real>>("neutron_energy_limits")),
+    _recoil_energy_limits(getParam<std::vector<Real>>("recoil_energy_limits")),
+    _L(getParam<unsigned int>("legendre_order"))
+{
+   /// See "enum_order.h and enum_quadrature_type.h"
+  _quadrature = QBase::build(QGAUSS, 1, FORTYTHIRD).release();
+
+  const std::vector<Point> & temp_points = _quadrature->get_points();
+  for (auto & pt : temp_points)
+  {
+    _quad_points.push_back(pt(0));
+    _console << pt(0) << std::endl;
+  }
+  _quad_weights = _quadrature->get_weights();
+
+
+}
+
+Real
+ElasticRecoilCrossSectionUserObject::legendreP(unsigned int  n, Real x)
+{
+  switch (n)
+  {
+    case 0:
+      return 1;
+
+    case 1:
+      return x;
+
+    case 2:
+      return 0.5 * (3 * pow(x, 2) - 1);
+
+    case 3:
+      return  0.5 * (5 * pow(x, 3) - 3 * x);
+
+    case 4:
+      return 0.125 * (35 * pow(x, 4) - 30 * pow(x, 2) + 3);
+
+    case 5:
+      return 0.125 * (63 * pow(x, 5) - 70 * pow(x, 3) + 15 * x);
+
+    default:
+      mooseError("Implementation of Legendre polynomials goes up to n = 5");
+  }
+}
+
+/// Class destructor
+ElasticRecoilCrossSectionUserObject::~ElasticRecoilCrossSectionUserObject()
+{
+  delete _quadrature;
+}
+
+void
+ElasticRecoilCrossSectionUserObject::initialize()
+{
+  /// Loop to calculate neutron spectrum over group g xi(E)
+  unsigned int G = _neutron_energy_limits.size() - 1;
+  _xi_g.resize(G);
+  for (unsigned int g = 0; g < G; ++g)
+  {
+    Real E_max = _neutron_energy_limits[g];
+    Real E_min = _neutron_energy_limits[g + 1];
+
+    unsigned int nqp = _quad_points.size();
+
+    for (unsigned int p = 0; p < nqp; ++p)
+    {
+      Real E = 0.5 * (E_max - E_min) * _quad_points[p] + 0.5 * (E_max + E_min);
+      Real w_E = 0.5 * _quad_weights[p] * (E_max - E_min);
+
+      _xi_g[g] += w_E * _neutron_spectrum.value(E, Point());
+    }
+  }
+}
+
+/// Find the neutron energy group
+unsigned int
+ElasticRecoilCrossSectionUserObject::findNeutronEnergyGroup(Real energy)
+{
+  unsigned int G = _neutron_energy_limits.size() - 1;
+  Real E_max = _neutron_energy_limits[0];
+  unsigned int group;
+
+  for (unsigned int g = 0; g < G; ++g)
+  {
+    if (energy > E_max)
+    {
+      group = 1;
+    }
+    else if (energy > _neutron_energy_limits[g+1] && energy < _neutron_energy_limits[g])
+    {
+      group = g + 1;
+    }
+    else
+    {
+      group = G;
+    }
+  }
+  return group;
+}
+
+void
+ElasticRecoilCrossSectionUserObject::execute()
+{
+  _gamma = 4 * _atomic_mass[0] / std::pow(( _atomic_mass[0] + 1),2);
```

Comment:
`_gamma`  does not change except when changing `_atomic_mass`. I think we want one UserObject for 
each nuclide. Hence `_atomic_mass` should be a single Real number. Then `_gamma` can be computed in the constructor.